### ISSUE TITLE
Windows+Mingw+Codeblocks failing

### ISSUE THREAD

**Issue by koraa** [inexor-game/code#121]
This still seems to be an issue.

@IAmNotHanni Could you please, at your convenience, elaborate on the state of this?

**Comment by koraa** [comment]
@IAmNotHanni Please provide more info.

**Comment by Croydon** [comment]
Here once again the logs: https://gist.github.com/Croydon/1206b368550d2c9ecd74

A CmakeError file isn't getting created. Also yes, the submodules are up to date, however note that I'm using this commit: https://github.com/inexor-game/platform_windows/commit/c870e58bf907c69e1bba832394c2f04f11716b3a since it seems more correct, also without this, there aren't any useful information in the build log, at least from what I saw, but perhaps I'm wrong.

**Comment by koraa** [comment]
Fantastic, thanks!

Please use platform_windows from master.

Could you also post CMakeFiles\inexor.dir\linklibs.rsp ? It contains the libraries that are actually linked.
It's being inserted into the command line.

**Comment by Croydon** [comment]
build\src\client\CMakeFiles\inexor.dir: https://gist.github.com/Croydon/c7b97fc9a99587017b1b

**Comment by koraa** [comment]
compairing your linklibs.rsp and the logs above, it seems you're linking 32 bit libs but compiling for 64 bit.
Please try setting MINGW_TYPE=x86_64

**Comment by koraa** [comment]
src/platform_windows/mingw.cmake:6 contains code that tries to guess the architecture. @a-teammate Would that override setting -DMINGW_TYPE ?

**Comment by a-teammate** [comment]
I think so yes.

**Comment by koraa** [comment]
Could you please patch it not to do that?
you could probably use `option`

**Comment by Efreak** [comment]
I am attempting using your [build instructions](https://github.com/inexor-game/code/wiki/Build#windows), and having issues generating the makefiles with CMake (GUI if it matters, though it shouldn't). When I select *CodeBlocks - MinGW Makefiles*, I get an *libiconv-2.dll is missing* error 5 times, then I get an *error in configuration process*. *CMAKE_BUILD_TYPE* is blank, and *CMAKE_CODEBLOCKS_EXECUTABLE*'s value is *CMAKE_CODEBLOCKS_EXECUTABLE-NOTFOUND*

[![CMake & Errors](http://i.imgur.com/hnuwPgrl.png)](http://i.imgur.com/hnuwPgr.png)
[Full CMake error log](http://pastebin.com/aCtQ83v6)

Also I should note that while GFW is listed as an alternative, it says to install SmartGit, which doesn't actually install any git binaries...

**Comment by koraa** [comment]
@Efreak Where is your code blocks installed? I think CMake expects it to be under  `C:/Program Files (x86)/CodeBlocks/MinGW/bin/gcc.exe`

**Comment by koraa** [comment]
@Efreak Also, feel free to post the cygwin errors in a separate Issue! The more data we have the better.

**Comment by Croydon** [comment]
The CMake errors from the log of @Efreak seems to be identical with mine, when I'm setting MINGW_TYPE=x86_64
Note: When I just use all default options, I can generate the Code::Blocks project file, but then it's failing at compiling, like posted before.

**Comment by koraa** [comment]
Well, does either of you have code blocks installed at  C:/Program Files (x86)/CodeBlocks ?

**Comment by Croydon** [comment]
The path CMake detects is correct, that's not the problem. If you need to know if for something else, yes, CodeBlocks is there, but I have installed a newever version of MinGW seperated.

**Comment by koraa** [comment]
could you please investigate that? Given that I am on linux I really have a hard time debugging that stuff.

**Comment by Efreak** [comment]
I am unable to check on this until wednesday at the earliest, as I am working on this using an older computer at school. I am unsure if it is 32bit or 64bit, though looking at @Croyden's comment regarding I do know that I installed CMake and Code::Blocks both to the default folders. I will test again, however, and see if I get the same error at home.


From my logfile, I'm guessing that the Code::Blocks install does exist in C:\program files (x86)\, since the error is an "Interrupt/exception caught," not a missing binary.

**Comment by Efreak** [comment]
I can verify that libiconv is installed in the correct location.

[![libiconv-2.dll](http://i.imgur.com/zQYuLLis.png)](http://i.imgur.com/zQYuLLi.png) [![All relevant files](http://i.imgur.com/7AuB9ib.png)](http://i.imgur.com/7AuB9ib.png)

**Comment by koraa** [comment]
The warnings should be fixed now. The rest sounds like problems building rpc. Can you check if your platform is capable of executing src/platform_windows/tool/windows/protoc.exe ?

I have no Idea about the IAR error, but a quick research yielded this:
* https://stackoverflow.com/questions/26885865/clion-with-mingw-not-compiling-simple-hello-world-app

That sort of sounds like a dependency problem.

**Comment by Efreak** [comment]
````
CIT-M27MF6J49GT+Administrator@CIT-M27MF6J49GT /cygdrive/c/Users/Administrator/Desktop/inexor
$ src/platform_windows/tool/windows/protoc.exe
Missing input file.
````

**Comment by a-teammate** [comment]
what has worked for me was:
* install http://tdm-gcc.tdragon.net/download (64 bit)
* install codeblocks without mingw
* run codeblocks and create a simple test programm to verify its compiling.
   * i had to do that because it tried for mingw32-gcc.exe instead of gcc.exe (Settings -> compiler -> toolchain executables, (dunno whether you need that though) (and ofc the others as well)
* **Make Sure you got the newest master**
   * It has updated, click on 'pull' in smartgit or 'git pull' on the commandline

* use codeblocks - mingw makefiles for the cmake generator, dont forget to delete the cache before
* open the created codeblocks file
* only build 'inexor' (instead of 'all')

**Comment by Croydon** [comment]
My issues are now completely fixed. I also don't get any error when I build "all".

**Comment by a-teammate** [comment]
oh yeah i see, my "build" folder was contaminated. deleting it and rerunning cmake did it.

**Comment by Efreak** [comment]
I actually found the issue when I installed MinGW and had similar errors as those for CodeBlocks, though for different libraries (libgmp). MinGW installer puts libgmp and other dll files in c:\mingw\bin. However, it doesn't add the C:\mingw paths to the system path. You need to do this manually. 

Code::Blocks also does this. However, it puts its MinGW install in `c:\Program Files (x86)\CodeBlocks\MinGW\bin`, not in `C:\MinGW`, so you need to add `c:\Program Files (x86)\CodeBlocks\MinGW\bin` to your system path (open `rundll32 sysdm.cpl,EditEnvironmentVariables` from the run menu). CMake runs fine now. I just realized, however, that for some reason I'm on the cmake-mingw-fix branch. I'll switch over and try again, and also try compiling.

I would imagine that for general use, the Code::Blocks people never thought to add their MinGW to the system path, as it's not there for system use.

**Comment by Croydon** [comment]
Even before I installed TDM-GCC (to get a newer MinGW version), I had no troubles to compile Inexor. I just used the default stable release of Code::Blocks with MinGW. I can't say much more to this, but I guess the error is somewhere else.

**Comment by Croydon** [comment]
@Efreak Do you still experience troubles with the master branch?

**Comment by Efreak** [comment]
Yes, I'm still trying to figure out what it is. I'm going to try installing TDM-GCC separately as soon as I can, but I don't think my choice of compiler is the issue, I think it's the code itself, but I'm not clear on whether the issues are the same with CB and VS.

I just pulled updates and tried compiling again about 10 minutes ago, so all errors are as of the source tree *now*. Prior to doing this, I wiped all the code off my computer, and regenerated the projects using cmake.

I have created four separate folders for this:

* `D:\inexor` - the repository
* `D:\inexor_codeblocks` - uses CodeBlocks MinGW project in build folder
* `D:\inexor_vcx86` - uses Visual Studio 2013 project in build folder
* `D:\inexor_vcx64` - uses Visual Studio 2013 Win64 project in build folder

#### Code::Blocks:
* I have reinstalled Code::Blocks from scratch
* I get the [same errors](http://i.imgur.com/KZHPEhw.png) there as before
* I've never used Code::Blocks for anything in the past, so I don't really know where it's supposed to put binaries, however it doesn't create any new folders...

#### Visual Studio 2013 with regular CMake project
* Debug
    * Compiles with only a [warning](http://i.imgur.com/mH06nvC.png).
    * No binaries in the usual folder, `Win32/Debug`.
    * Creates an empty *build/Debug* folder
* Release
    * Compiles with no errors.
    * No binaries in the usual folder `x64\Release`
    * Creates an empty *build/Release* folder

#### Visual Studio 2013 with Win64 CMake project
* Debug:
    * Compiles with only a [warning](http://i.imgur.com/nfv1T9Q.png).
    * No binaries in the folder, `x64\Debug`
* Release
    * Compiles with no errors.
    * No binaries in the folder `x64\Release`


In all four cases for Visual Studio, [there are several executable files that get created](http://i.imgur.com/Jk6cv0v.png), however none of them get moved into the release folder. I do not know what the end-result folder structure is supposed to be. inexor.exe crashes from a missing SDL2.dll, which is in the source tree.

-----

If you want, I have uploaded the relevant data [to my dropbox](https://dl.dropboxusercontent.com/u/10951211/temp/buildfolders.txz). This is the contents of all build folders, *except* the build/src folder (~180mb) and build/inexor.sdf file (~85mb). If you need any other information from me, I'm happy to provide it if I can.

**Comment by a-teammate** [comment]
according to this: http://stackoverflow.com/questions/6312151/stricmp-with-mingw-and-c0x-not-existent
it cant find _stricmp because we use standard c++11 and _stricmp is not ansi standard. Sidenote: _stricmp has already been in sauerbraten before, we just upgraded the language standard.
(Actually i think this problem could disappear after upgrading your compiler to the tdm one, still we should fix that)

To your issues with Visual Studio: after "compiling" the "install" project in your solution explorer, the binaries will go to `bin/windows/win{32, 64}/`

P.S: You do not need to clear your whole repository usually when building other project files with cmake.  
Though you probably need to generate those projects into another directory, or delete the `build` directory after switching the generator (e.g. after switching from visual studio to mingw)

**Comment by Efreak** [comment]
>according to this: http://stackoverflow.com/questions/6312151/stricmp-with-mingw-and-c0x-not-existent
it cant find _stricmp because we use standard c++11 and _stricmp is not ansi standard. Sidenote: _stricmp has already been in sauerbraten before, we just upgraded the language standard.
(Actually i think this problem could disappear after upgrading your compiler to the tdm one, still we should fix that)

Go figure I didn't even look for any solutions specific to MinGW. I spent a lot of time looking up stricmp and _stricmp and the other functions. I did find that information about C++11, however, and I tried enabling this manually in Code::Blocks (which I should note *does* have TDM-GCC in it, it's just the one it comes with, not a separate version).

I've compiled the project, including the install project. This seems like it would have been a problem from the beginning; there is no mention anywhere of building for for the 'INSTALL' project on the [Build](https://github.com/inexor-game/code/wiki/Build#windows) page of the wiki (well, there will be in a couple minutes, I'll add it myself...).

I am currently downloading the data; if anything goes wrong with running the game once I have the data, I will let you know; otherwise this is all fixed.
